tests/dir_helpers.py

```diff
@@ -0,0 +1,229 @@
+from __future__ import unicode_literals
+
+import os
+import pytest
+
+from funcy.py3 import lmap, retry
+
+from dvc.utils import makedirs
+from dvc.utils.compat import basestring, is_py2, pathlib, fspath, fspath_py35
+
+
+__all__ = ["tmp_dir", "scm", "dvc", "repo_template", "run_copy", "erepo_dir"]
+REPO_TEMPLATE = {
+    "foo": "foo",
+    "bar": "bar",
+    "dir": {
+        "data": "dir/data text",
+        "subdir": {"subdata": "dir/subdir/subdata text"},
+    },
+}
+
+
+class TmpDir(pathlib.Path):
+    def __new__(cls, *args, **kwargs):
+        if cls is TmpDir:
+            cls = WindowsTmpDir if os.name == "nt" else PosixTmpDir
+        self = cls._from_parts(args, init=False)
+        if not self._flavour.is_supported:
+            raise NotImplementedError(
+                "cannot instantiate %r on your system" % (cls.__name__,)
+            )
+        self._init()
+        return self
+
+    # Not needed in Python 3.6+
+    def __fspath__(self):
+        return str(self)
+
+    def _require(self, name):
+        if not hasattr(self, name):
+            raise TypeError(
+                "Can't use {name} for this temporary dir. "
+                'Did you forget to use "{name}" fixture?'.format(name=name)
+            )
+
+    def gen(self, struct, text=""):
```

Comment:
really neat; hard to read at the beginning but the interface is awesome :heart_eyes: 
would you mind documenting the intended usage, @Suor? (a quick summary on top of the module is enough))

Reply:
This is a good idea.